that_depends/providers/base.py

```diff
@@ -125,35 +125,34 @@ def sync_tear_down(self) -> None:
             raise RuntimeError(msg)
 
 
+ResourceCreator: typing.TypeAlias = (
+    typing.Callable[
+        P,
+        typing.Iterator[T_co]
+        | typing.AsyncIterator[T_co]
+        | typing.ContextManager[T_co]
+        | typing.AsyncContextManager[T_co],
+    ]
+    | typing.ContextManager[T_co]
+    | typing.AsyncContextManager[T_co]
+)
+
+
 class AbstractResource(AbstractProvider[T_co], abc.ABC):
     def __init__(
         self,
-        creator: typing.Callable[P, typing.Iterator[T_co] | typing.AsyncIterator[T_co]],
+        creator: ResourceCreator[P, T_co],
         *args: P.args,
         **kwargs: P.kwargs,
     ) -> None:
         super().__init__()
-        if inspect.isasyncgenfunction(creator):
-            self._is_async = True
-        elif inspect.isgeneratorfunction(creator):
-            self._is_async = False
-        else:
-            msg = f"{type(self).__name__} must be generator function"
-            raise RuntimeError(msg)
-
-        self._creator: typing.Final = creator
-        self._args: typing.Final = args
-        self._kwargs: typing.Final = kwargs
-
-    def _is_creator_async(
-        self, _: typing.Callable[P, typing.Iterator[T_co] | typing.AsyncIterator[T_co]]
-    ) -> typing.TypeGuard[typing.Callable[P, typing.AsyncIterator[T_co]]]:
-        return self._is_async
-
-    def _is_creator_sync(
-        self, _: typing.Callable[P, typing.Iterator[T_co] | typing.AsyncIterator[T_co]]
-    ) -> typing.TypeGuard[typing.Callable[P, typing.Iterator[T_co]]]:
-        return not self._is_async
+        is_async, normalized_creator = _normalize_creator(self, creator, *args, **kwargs)
+        self._is_async = is_async
```

Comment:
not sure if it's necessary

Reply:
I'm not sure if it's useful to keep `is_async` in abstract resource.

`ResourceContext` instance holds CM stack (it may be a `AsyncExitStack` instance on `async_resolve` call  or `ExitStack` on `sync_resolve`). Then the tear down process invokes async `aclose` or `close` accordingly, based on CM stack `isinstance` check. Thus, `ResourceContext` does not really need `is_async` attribute and `AbstractResource` either.

Reply:
here, it checks if resource can be torn down https://github.com/modern-python/that-depends/blob/main/that_depends/providers/base.py#L170

so, if resource is async and context is sync, than resolving is forbidden

Reply:
In my PR this block of code was removed. I didn't find a case when context is not asynchronous but creator is asynchronous. :thinking: